### PR TITLE
refactor: ids and licenses

### PR DESCRIPTION
**Pull request by mockdot**
## Description

- IDs are now given from qbx-cityhall using um-idcard
- giveStarterItems function is now only used to give starter items defined in QBCore.Shared.StarterItems

## Checklist

- [ ] I have personally loaded this code into an updated Qbox project and checked all of its functionality.
- [x] My code fits the style guidelines.
- [x] My PR fits the contribution guidelines.

### CODE REVIEW

File: client/character.lua [hunk 1]
```diff
@@ -212,6 +212,7 @@ local function createCharacter(cid)
         cid = cid
     })
     destroyPreviewCam()
+    TriggerServerEvent('qb-cityhall:server:getIDs')
```
Comment (Manason) [inline client/character.lua]: Is this needed? I thought qbx-cityhall was consuming the character creation event to give ids?

File: shared/main.lua [hunk 2]
```diff
@@ -2,8 +2,6 @@ QBShared = QBShared or {}
 
 QBShared.StarterItems = {
     phone = { amount = 1, item = 'phone' },
-    id_card = { amount = 1, item = 'id_card' },
-    driver_license = { amount = 1, item = 'driver_license' },
 }
```
Comment (Manason) [inline shared/main.lua:5]: Should we move this to config? It doesn't really seem appropriate as part of QBShared

**Review by Manason — CHANGES_REQUESTED** [review]

**Review by Manason — APPROVED** [review]

File: shared/main.lua (hunk 2 above)
Comment (mockdot) [reply shared/main.lua:5]: I think that's a good idea, let me go ahead and move it

**Comment by mockdot** [comment]
I have not been able to test it yet, getting black screen when creating chars, but trying to show the idea

File: config.lua [hunk 3]
```diff
@@ -132,3 +129,9 @@ Config.CharacterDataTables = {
     ['player_outfits'] = 'citizenid',
     ['player_vehicles'] = 'citizenid',
 } -- Rows to be deleted when the character is deleted
+
+Config.StarterItems = { -- Character starting items
+    phone = { amount = 1, item = 'phone' },
+    id_card = { amount = 1, item = 'id_card', type = 'id'},
+    driver_license = { amount = 1, item = 'driver_license', type = 'license'},
+}
```
Comment (Manason) [inline config.lua:137]: This should be an array

File: config.lua [hunk 4]
```diff
@@ -132,3 +129,9 @@ Config.CharacterDataTables = {
     ['player_outfits'] = 'citizenid',
     ['player_vehicles'] = 'citizenid',
 } -- Rows to be deleted when the character is deleted
+
+Config.StarterItems = { -- Character starting items
+    phone = { amount = 1, item = 'phone' },
+    id_card = { amount = 1, item = 'id_card', type = 'id'},
+    driver_license = { amount = 1, item = 'driver_license', type = 'license'},
```
Comment (Manason) [inline config.lua:136]: Needs a field for metadata? What is type?

File: server/character.lua [hunk 5]
```diff
@@ -6,10 +6,53 @@ local function getAllowedAmountOfCharacters(license2, license)
...
+    for _, v in pairs(Config.StarterItems) do
+        if v.type == 'id' then
+            local metadata = {
+                cardtype = v,
+                citizenid = Player.PlayerData.citizenid,
+                firstname = Player.PlayerData.charinfo.firstname,
+                lastname = Player.PlayerData.charinfo.lastname,
+                birthdate = Player.PlayerData.charinfo.birthdate,
+                sex =  getSexString(Player.PlayerData.charinfo.gender),
+                nationality = Player.PlayerData.charinfo.nationality,
```
Comment (Manason) [inline server/character.lua:33]: metadata should be in the config itself, even if that means putting a function in the config

**Review by Manason — CHANGES_REQUESTED** [review]

**Comment by Manason** [comment]
Where is money added? We should probably move that to this as well since it's an item in ox_inventory

File: config.lua [hunk 6]
```diff
@@ -129,3 +129,55 @@ Config.CharacterDataTables = {
...
+        nationality = PlayerData.charinfo.nationality,
+        mugShot = 'none',
+    }
+    return metadata
+end
+
+---@alias ItemName string
+---@alias ItemAmount number
+---@alias ItemMetadata table
+---@type table<ItemName, ItemAmount, ItemMetadata>
```
Comment (Manason) [inline config.lua:178]:
```suggestion
---@type { item: string, amount: integer, metadata: table|function }
```
I don't believe you can specify a table<> with more than two values

File: config.lua [hunk 7]
```diff
@@ -129,3 +129,55 @@ Config.CharacterDataTables = {
...
+Config.StarterItems = { -- Character starting items
+    { item = 'phone', amount = 1 },
+    { item = 'id_card', amount = 1, metadata = getItemMetadata('id_card', 'id')},
```
Comment (Manason) [inline config.lua:181]: This needs to be an in-line function, as it cannot be evaluated when the file loads, but instead needs to be evaluated when the item is being added to the character.

**Review by Manason — CHANGES_REQUESTED** [review]

File: config.lua (hunk 7 above)
Comment (mockdot) [reply config.lua:181]: You are referring to using getItemMetadata()? I should specify the metadata directly instead of using the function?

**Comment by mockdot** [comment]
Seems to work great!

File: config.lua [hunk 8]
```diff
@@ -129,3 +129,10 @@ Config.CharacterDataTables = {
     ['player_outfits'] = 'citizenid',
     ['player_vehicles'] = 'citizenid',
 } -- Rows to be deleted when the character is deleted
+
+---@type { name: string, amount: integer, metadata: table | fun(source: number): table}
+Config.StarterItems = { -- Character starting items
+    { name = 'phone', amount = 1 },
+    { name = 'id_card', amount = 1, metadata = function(source) exports['um-idcard']:GetMetaLicense(source, {'id_card'}) end },
+    { name = 'driver_license', amount = 1, metadata = function(source) exports['um-idcard']:GetMetaLicense(source, {'driver_license'}) end },
```
Comment (BerkieBb) [inline config.lua:137]: I think you forgot the `return` in both functions

**Review by BerkieBb — CHANGES_REQUESTED** [review]

**Comment by Manason** [comment]
Awaiting new release of um-idcard to approve this. We should also add a lib.checkVersion within the metadata function to provide extra verification that people are running um-idcard

File: config.lua [hunk 9]
```diff
@@ -129,3 +129,10 @@ Config.CharacterDataTables = {
     ['player_outfits'] = 'citizenid',
     ['player_vehicles'] = 'citizenid',
 } -- Rows to be deleted when the character is deleted
+
+---@type { name: string, amount: integer, metadata: table | fun(source: number): table}
+Config.StarterItems = { -- Character starting items
+    { name = 'phone', amount = 1 },
+    { name = 'id_card', amount = 1, metadata = function(source) if not lib.checkDependency('um-idcard', '1.2.8') then return else exports['um-idcard']:GetMetaLicense(source, {'id_card'}) end end },
```
Comment (Manason) [inline config.lua:136]:
```suggestion
    { name = 'id_card', amount = 1, metadata = function(source) if not lib.checkDependency('um-idcard', '1.2.8', true) then error() end exports['um-idcard']:GetMetaLicense(source, {'id_card'}) end end },
```
Prints and errors the dependency. We want to fail loud here.

**Review by Manason — CHANGES_REQUESTED** [review]

File: config.lua [hunk 10]
```diff
@@ -129,3 +129,10 @@ Config.CharacterDataTables = {
     ['player_outfits'] = 'citizenid',
     ['player_vehicles'] = 'citizenid',
 } -- Rows to be deleted when the character is deleted
+
+---@type { name: string, amount: integer, metadata: table | fun(source: number): table}
+Config.StarterItems = { -- Character starting items
+    { name = 'phone', amount = 1 },
+    { name = 'id_card', amount = 1, metadata = function(source) if not lib.checkDependency('um-idcard', '1.2.8', true) then error() end exports['um-idcard']:GetMetaLicense(source, {'id_card'}) end end },
+    { name = 'driver_license', amount = 1, metadata = function(source) if not lib.checkDependency('um-idcard', '1.2.8') then return else exports['um-idcard']:GetMetaLicense(source, {'driver_license'}) end end },
```
Comment (Manason) [inline config.lua:137]:
```suggestion
    { name = 'driver_license', amount = 1, metadata = function(source) if not lib.checkDependency('um-idcard', '1.2.8', true) then error() end exports['um-idcard']:GetMetaLicense(source, {'driver_license'}) end end },
```

**Review by Manason — CHANGES_REQUESTED** [review]

**Pull request merged**